network/connecteventmanager.go

```diff
@@ -0,0 +1,92 @@
+package network
+
+import (
+	"sync"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type ConnectionListener interface {
+	PeerConnected(peer.ID)
+	PeerDisconnected(peer.ID)
+}
+
+type connectEventManager struct {
+	connListener ConnectionListener
+	lk           sync.Mutex
+	conns        map[peer.ID]*connState
+}
+
+type connState struct {
+	refs       int
+	responsive bool
+}
+
+func newConnectEventManager(connListener ConnectionListener) *connectEventManager {
+	return &connectEventManager{
+		connListener: connListener,
+		conns:        make(map[peer.ID]*connState),
+	}
+}
+
+func (c *connectEventManager) Connected(p peer.ID) {
+	c.lk.Lock()
+	defer c.lk.Unlock()
+
+	state, ok := c.conns[p]
+	if !ok {
+		state = &connState{responsive: true}
+		c.conns[p] = state
+	}
+	state.refs++
+
+	if state.refs == 1 && state.responsive {
```

Comment:
I'd consider switching the peer back to "responsive" on connect.

Reply:
That is:

```go
if state.refs == 1 || !state.responsive {
    state.responsive = true
    ...
}
```

Reply:
We get into the unresponsive state if the remote peer fails to respond to several attempts to dial it, but it's still connected. So we're implicitly saying that we care about responsiveness more than connectivity.
I guess arguably if a peer opens a new connection it can be considered responsive?

Reply:
Thinking about this more we probably want to keep it how it is - if for example a peer that doesn't support bitswap dials us, we will dial it when broadcasting to connected peers. If the peer responds with an error indicating protocol not supported, we shouldn't try to dial it again even if it connects to us again.

Reply:
Given that we're making 3 attempts, it's probably fine. I'm concerned that it can sometimes take some time to know that a connection is _actually_ dead. In that case, we could try several times, say "peer's dead!", then get the new connection, then see the old connection finally die.